src/funnel.test.ts

```diff
@@ -0,0 +1,1049 @@
+
+      await sleep(0);
+
+      expect(mockFn).toHaveBeenCalledTimes(1);
+      expect(mockFn).toHaveBeenLastCalledWith(["a", "b", "c"]);
+
+      foo.call("d");
+      foo.call("e");
+      foo.call("f");
+
+      expect(mockFn).toHaveBeenCalledTimes(1);
+
+      await sleep(0);
+
+      expect(mockFn).toHaveBeenCalledTimes(2);
+      expect(mockFn).toHaveBeenLastCalledWith(["d", "e", "f"]);
+    });
+
+    test("with a non-trivial delay timer", async () => {
+      const mockFn = vi.fn();
+      const foo = funnel(mockFn, {
+        reducer: ARGS_COLLECTOR,
+        invokedAt: "both",
+        delayMs: UT,
+        burstCoolDownMs: 0,
+      });
+      foo.call("a");
+      foo.call("b");
+
+      expect(mockFn).toHaveBeenCalledTimes(1);
+      expect(mockFn).toHaveBeenLastCalledWith(["a"]);
+
+      await sleep(0);
+
+      expect(mockFn).toHaveBeenCalledTimes(1);
+
+      await sleep(UT);
+
+      expect(mockFn).toHaveBeenCalledTimes(2);
+      expect(mockFn).toHaveBeenLastCalledWith(["b"]);
+    });
+
+    test("burst timer with non-trivial maxBurstDuration", async () => {
+      const mockFn = vi.fn();
+      const foo = funnel(mockFn, {
+        reducer: ARGS_COLLECTOR,
+        burstCoolDownMs: 0,
+        maxBurstDurationMs: UT,
+        invokedAt: "end",
+      });
+      foo.call("a");
+      foo.call("b");
+      foo.call("c");
+
+      expect(mockFn).toHaveBeenCalledTimes(0);
+
+      await sleep(0);
+
+      expect(mockFn).toHaveBeenCalledTimes(1);
+      expect(mockFn).toHaveBeenLastCalledWith(["a", "b", "c"]);
+    });
+  });
+
+  test("both timers", async () => {
+    const mockFn = vi.fn();
+    const foo = funnel(mockFn, {
+      reducer: ARGS_COLLECTOR,
+      invokedAt: "both",
+      delayMs: 0,
+      burstCoolDownMs: 0,
+    });
+    foo.call("a");
+
+    expect(mockFn).toHaveBeenCalledTimes(1);
+    expect(mockFn).toHaveBeenLastCalledWith(["a"]);
+
+    await sleep(0);
+
+    expect(mockFn).toHaveBeenCalledTimes(1);
+  });
+
+  test("maxBurstDurationMs = 0 limits the burst immediately", async () => {
+    const mockFn = vi.fn();
+    const foo = funnel(mockFn, {
+      reducer: ARGS_COLLECTOR,
+      invokedAt: "end",
+      burstCoolDownMs: UT,
+      maxBurstDurationMs: 0,
+    });
+    foo.call("a");
+    await sleep(0);
+
+    expect(mockFn).toHaveBeenCalledTimes(1);
+    expect(mockFn).toHaveBeenLastCalledWith(["a"]);
+  });
+
+  test("all timeouts zero", async () => {
+    const mockFn = vi.fn();
+    const foo = funnel(mockFn, {
+      reducer: ARGS_COLLECTOR,
+      invokedAt: "both",
+      delayMs: 0,
+      burstCoolDownMs: 0,
+      maxBurstDurationMs: 0,
+    });
+    foo.call("a");
+    foo.call("b");
+
+    expect(mockFn).toHaveBeenCalledTimes(1);
+    expect(mockFn).toHaveBeenLastCalledWith(["a"]);
+
+    await sleep(0);
+
+    expect(mockFn).toHaveBeenCalledTimes(2);
+    expect(mockFn).toHaveBeenLastCalledWith(["b"]);
+  });
+});
+
+describe("disabled timers (no defined timeout durations)", () => {
```

Comment:
**question:** do you think we should support these cases?

Reply:
We can default the burst timer to 0 when nothing is provided; it's probably the most sensible option.

Reply:
i think that makes sense, we can also forbid it at the type-level?